lib/utilities/clean-remove.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+var path = require('path');
+var fs = require('fs-extra');
+var Promise = require('../ext/promise');
+var walkUp = require('./walk-up-path');
+var stat = Promise.denodeify(fs.stat);
+var remove = Promise.denodeify(fs.remove);
+var readdir = Promise.denodeify(fs.readdir);
+
+function cleanRemove(fileInfo) {
+  // see if file exists to avoid wasting time
+  return stat(fileInfo.outputPath).then(function() {
+    return remove(fileInfo.outputPath).then(function() {
+      var paths = walkUp(fileInfo.displayPath).map(function(thePath) {
+        return path.join(fileInfo.outputBasePath, thePath);
+      });
+
+      var linkedPromise = paths.reduce(function(linkedPromise, thePath) {
```

Comment:
Can you name the two `linkedPromise`s something a bit more clear? Maybe `completionPromise` and `chainedPromise`? Also add a note that you're using `reduce` to continually attach these head to tail. (The fact that it is shadowed makes it a bit confusing.)